File: Lecture-8-Build-Netflix-Recommendation-System/code/goDemo/internal/controller/movie_controller.go
package controller

import (
	"encoding/json"
	"errors"
	"net/http"

	"goDemo/internal/service"
)

type MovieController struct {
	movieService *service.MovieService
}

func NewMovieController(movieService *service.MovieService) *MovieController {
	return &MovieController{movieService: movieService}
}

// RegisterRoutes wires the /movies endpoints onto mux.
func (c *MovieController) RegisterRoutes(mux *http.ServeMux) {
	mux.HandleFunc("GET /movies/search", c.search)
	mux.HandleFunc("GET /movies/{title}/similar", c.similarMovies)
}

func (c *MovieController) search(w http.ResponseWriter, r *http.Request) {
	query := r.URL.Query().Get("query")

	matches, err := c.movieService.Search(r.Context(), query)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, matches)
}

func (c *MovieController) similarMovies(w http.ResponseWriter, r *http.Request) {
	title := r.PathValue("title")

	matches, err := c.movieService.SimilarMovies(title)
	if err != nil {
		if errors.Is(err, service.ErrMovieNotFound) {
			writeError(w, http.StatusNotFound, err.Error())
			return
		}
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, matches)
}

func writeJSON(w http.ResponseWriter, status int, payload any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(payload)
}

func writeError(w http.ResponseWriter, status int, message string) {
	writeJSON(w, status, map[string]string{"detail": message})
}

File: Lecture-8-Build-Netflix-Recommendation-System/code/pythonDemo/src/pythondemo/movie_controller.py
from fastapi import APIRouter, HTTPException

from .model.movie_match import MovieMatch
from .movie_service import MovieService


def create_movie_router(movie_service: MovieService) -> APIRouter:
    router = APIRouter(prefix="/movies", tags=["movies"])

    @router.get("/search", response_model=list[MovieMatch])
    def search(query: str) -> list[MovieMatch]:
        return movie_service.search(query)

    @router.get("/{title}/similar", response_model=list[MovieMatch])
    def similar_movies(title: str) -> list[MovieMatch]:
        try:
            return movie_service.similar_movies(title)
        except ValueError as error:
            raise HTTPException(status_code=404, detail=str(error)) from error

    return router
